File: apps/android/src/test/java/com/quran/android/telemetry/TelemetrySpy.kt
package com.quran.android.telemetry

import com.quran.android.app.AppError

/**
 * Test double for [Telemetry]. Records every call so unit tests can
 * assert event names, attribute keys, and error codes. Mirrors
 * `apps/ios/Tests/QuranRecitationAppTests/TelemetrySpy.swift`.
 *
 * `measure` records `<name>.succeeded` / `<name>.failed` synthetic
 * events with a `duration_ms="0"` placeholder so assertions are
 * deterministic; tests should not rely on the value.
 */
class TelemetrySpy : Telemetry {
    sealed interface Record {
        data class Event(val name: String, val attributes: Map<String, String>) : Record
        data class Error(val code: String, val context: Map<String, String>) : Record
    }

    private val _records = mutableListOf<Record>()
    val records: List<Record> get() = _records.toList()

    val eventNames: List<String>
        get() = _records.filterIsInstance<Record.Event>().map { it.name }

    fun reset() {
        _records.clear()
    }

    override fun event(name: String, attributes: Map<String, String>) {
        _records += Record.Event(name, attributes)
    }

    override suspend fun <T> measure(name: String, block: suspend () -> T): T {
        return try {
            val result = block()
            _records += Record.Event("$name.succeeded", mapOf("duration_ms" to "0"))
            result
        } catch (cause: Throwable) {
            val attrs = buildMap {
                put("duration_ms", "0")
                if (cause is AppError) put("error_code", cause.telemetryCode)
            }
            _records += Record.Event("$name.failed", attrs)
            throw cause
        }
    }

    override fun error(error: AppError, context: Map<String, String>) {
        _records += Record.Error(error.telemetryCode, context)
    }
}

File: apps/ios/Tests/QuranRecitationAppTests/TelemetrySpy.swift
import Foundation
@testable import QuranRecitationApp

/// `Telemetry` test double that captures events and errors so tests can
/// assert on what was reported. `measure(_:_:)` runs the block and
/// records both a `<name>.measured` start marker and a final
/// `<name>.succeeded` / `.failed` event with `duration_ms` set to "0"
/// (deterministic for assertions).
final class TelemetrySpy: Telemetry {
  enum Record: Equatable {
    case event(name: String, attributes: [String: String])
    case error(code: String, context: [String: String])
  }

  private(set) var records: [Record] = []

  func event(_ name: String, attributes: [String: String]) {
    records.append(.event(name: name, attributes: attributes))
  }

  func measure<T>(_ name: String, _ block: () async throws -> T) async rethrows -> T {
    do {
      let result = try await block()
      records.append(.event(name: name + ".succeeded", attributes: ["duration_ms": "0"]))
      return result
    } catch {
      records.append(.event(name: name + ".failed", attributes: ["duration_ms": "0"]))
      throw error
    }
  }

  func error(_ error: AppError, context: [String: String]) {
    var combined = context
    combined["error_code"] = error.telemetryCode
    records.append(.error(code: error.telemetryCode, context: combined))
  }

  func eventNames() -> [String] {
    records.compactMap {
      if case let .event(name, _) = $0 { return name } else { return nil }
    }
  }
}
